cpp/src/io/convert/dlpack/cudf_dlpack.cpp

```diff
@@ -0,0 +1,247 @@
+                          int *num_columns,
+                          DLManagedTensor const * tensor)
+{
+  // We can copy from host or device pointers
+  GDF_REQUIRE(kDLGPU == tensor->dl_tensor.ctx.device_type ||
+              kDLCPU == tensor->dl_tensor.ctx.device_type ||
+              kDLCPUPinned == tensor->dl_tensor.ctx.device_type, 
+              GDF_INVALID_API_CALL);
+  // Make sure the current device ID matches the Tensor's device ID
+  int device_id = 0;
+  CUDA_TRY(cudaGetDevice(&device_id));
+  GDF_REQUIRE(tensor->dl_tensor.ctx.device_id == device_id, 
+              GDF_INVALID_API_CALL);
+
+  // Currently only 1D and 2D tensors are supported
+  GDF_REQUIRE(tensor->dl_tensor.ndim > 0, GDF_DATASET_EMPTY);
+  GDF_REQUIRE(tensor->dl_tensor.ndim <= 2, GDF_NOTIMPLEMENTED_ERROR);
+
+  // Ensure the column is not too big
+  GDF_REQUIRE(tensor->dl_tensor.shape[0] > 0, GDF_DATASET_EMPTY);
+  GDF_REQUIRE(tensor->dl_tensor.shape[0] < 
+              std::numeric_limits<gdf_size_type>::max(), 
+              GDF_COLUMN_SIZE_TOO_BIG);
+  
+  // compute the GDF datatype
+  gdf_dtype dtype = DLDataType_to_gdf_dtype(tensor->dl_tensor.dtype); 
+  GDF_REQUIRE(dtype != GDF_invalid, GDF_UNSUPPORTED_DTYPE);
+  
+  // Note: since cuDF has no way to register a deleter for a column, we
+  // have to make a copy of the data. In the future if we add the ability 
+  // to register deleters, we could register a wrapper around the 
+  // DLManagedTensor's deleter and just use the memory directly without copying
+  // (Note: this is only true for 1D tensors and column-major 2D tensors. Other
+  // layouts will require copying anyway.)
+
+  // compute the size and allocate data
+  *num_columns = 1;
+  if (tensor->dl_tensor.ndim == 2) *num_columns = tensor->dl_tensor.shape[1];
+  *columns = new gdf_column[*num_columns];
+  GDF_REQUIRE(*columns != nullptr, GDF_MEMORYMANAGER_ERROR);
+
+  gdf_size_type byte_width = gdf_dtype_size(dtype);
+  gdf_size_type length = tensor->dl_tensor.shape[0];
+  size_t bytes = length * byte_width;
+
+  int64_t col_stride = 0;
+  if (*num_columns > 1) {
+    col_stride = byte_width * length;
+    if (nullptr != tensor->dl_tensor.strides)
+      col_stride = tensor->dl_tensor.strides[1];
+  }
+
+  // copy the dl_tensor data
+  for (gdf_size_type c = 0; c < *num_columns; ++c) {
+    void *tensor_data = reinterpret_cast<void*>(
+      reinterpret_cast<uintptr_t>(tensor->dl_tensor.data) +
+      tensor->dl_tensor.byte_offset +
+      col_stride);
+
+    void* col_data = 0;
+    RMM_TRY(RMM_ALLOC(&col_data, bytes, 0));
+
+    CUDA_TRY(cudaMemcpy(col_data, tensor_data, bytes, cudaMemcpyDefault));
+
+    // construct column view
+    gdf_error status = gdf_column_view(&(*columns)[c], col_data,
+                                       nullptr, length, dtype);
+    GDF_REQUIRE(GDF_SUCCESS == status, status);
+  }
+
+  // Call the managed tensor's deleter since our "borrowing" is done
+  tensor->deleter(const_cast<DLManagedTensor*>(tensor));
```

Comment:
The documentation on this from dlpack is limited. I think it should really be called `release` rather than `deleter`. Here is what the doc comment for DLManagedTensor says:

```
/*!
 * \brief C Tensor object, manage memory of DLTensor. This data structure is
 *  intended to faciliate the borrowing of DLTensor by another framework. It is
 *  not meant to transfer the tensor. When the borrowing framework doesn't need
 *  the tensor, it should call the deleter to notify the host that the resource
 *  is no longer needed.
 */
```

https://github.com/dmlc/dlpack/blob/5c792cef3aee54ad8b7000111c9dc1797f327b59/include/dlpack/dlpack.h#L149-L155

Reply:
Ah, okay. Then I just misunderstood the semantics of the Tensor object here. It sounds like this object doesn't actually "own" any memory and calling it's `deleter` doesn't really delete anything important.

In that case, I think you're probably doing the right thing here.

Reply:
I guess in the future, our deleters (in `to_dlpack`), will probably decrement a refcount which *may* lead to deletion of the memory, or may not if a reference is still owned (say by the original column).

Reply:
I believe this is causing a double free segfault for me in working with CuPy:
```
#0  0x00007ffff6a1f512 in free () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007fff7dc92f5e in __pyx_f_4cupy_4core_6dlpack_deleter (__pyx_v_tensor=0x2b60c6c0) at cupy/core/dlpack.cpp:2778
#2  0x00007fff7dc949cb in __pyx_f_4cupy_4core_6dlpack_pycapsule_deleter (__pyx_v_dltensor=<optimized out>)
    at cupy/core/dlpack.cpp:2705
#3  0x00007ffff796d7df in capsule_dealloc (o=0x7fff7c705600) at Objects/capsule.c:261
...
```

Reply:
The function CuPy registers with DLPack definitely looks to free the memory:
```
https://github.com/cupy/cupy/pull/1082/files#diff-ddf01ff512087ef616db57ecab88c6aeR78
```